docs, net: Network binding plugin developer-guide (#10991)
kubevirt/kubevirt: docs/network-binding-plugin.md

### Network Binding
The network binding defines how the domain (VM) network interface is wired
in the VM pod through the domain to the guest.

The network binding includes:

- Domain vNIC configuration.
- Pod network configuration (optional).
- Services to deliver network details to the guest (optional).
  E.g. DHCP server to pass the IP configuration to the guest.

### Plugins
The network bindings have been part of Kubevirt core API and codebase.
With the increase of the number of network bindings added and
frequent requests to tweak and change the existing network bindings,
a decision has been made to create a network binding plugin infrastructure.

The plugin infrastructure provides means to compose a network binding plugin
and integrate it into Kubevirt in a modular manner.

Kubevirt is providing several network binding plugins as references.
The following plugins are available:

- [passt](https://kubevirt.io/user-guide/virtual_machines/net_binding_plugins/passt/) [v1.1.0]
- [macvtap](https://kubevirt.io/user-guide/virtual_machines/net_binding_plugins/macvtap/) [v1.1.1]
- [slirp](https://kubevirt.io/user-guide/virtual_machines/net_binding_plugins/slirp/) [v1.1.0]

> ## The Zero Code Plugin

Reviewer: It seems the example in this section is for a scenario where we want to run a custom CNI to plumb the network inside the container.
I think its worth to mention it at the beginning.
Also, personally I dont know about zero-code-plugin term, I dont think most CNIs would work out of the box, aren't they?

option which provides a pre-defined core Kubevirt method to attach an interface
to the domain.

The currently supported domain attachment type is `tap` (v1.1.1) which
builds a domain interface configuration that points to the tap/macvtap
existing interface.

Such a binding plugin assumes that the CNI used for the network connectivity
exposes in the pod a `tap` or `macvtap` (type) interface with a name corresponding
to the hashed network name:

- `pod<hash network name>` (or plain `eth0` for the primary network)
- `tap<hash network name>` (or plain `tap0` for the primary network)

Kubevirt virt-controller will define these names on the pod multus annotation,
therefore there is no special action needed from the plugin author except
choosing a CNI that confirms with the CNI standard spec.

For clarification, the CNI used for the connectivity is the one defined
in the `NetworkAttachmentDefinition` object referenced from the VM network spec.

### Configuration example

Given the following registered binding plugin:

```yaml
apiVersion: kubevirt.io/v1
kind: KubeVirt
metadata:
  name: kubevirt
  namespace: kubevirt
spec:
  configuration:
    network:

>       Binding:

Reviewer 2: Should this be lower case?

        my-zero-code-binding:
          domainAttachment: tap
```

The VM interface/network spec will include a ref to the binding plugin
and the `NetworkAttachmentDefinition` named `mynad`:

```yaml
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  name: myvm
spec:
  template:
    spec:
      domain:
        devices:
          interfaces:
            - name: mynetwork
              binding:
                name: my-zero-code-binding
      networks:
      - name: mynetwork
        multus:
          networkName: mmynad
```

> The `mynad` NetworkAttachmentDefinition defines `cni-name` as the CNI plugin

Reviewer 3: So maybe it worth clarifying this is not the binding plugin CNI.
e.g - Note: This is not the binding plugin CNI.

> The `mynad` NetworkAttachmentDefinition defines `cni-name` as the CNI plugin

Author: Done

binary:

```yaml
kind: NetworkAttachmentDefinition
apiVersion: k8s.cni.cncf.io/v1
metadata:
  name: mynad
spec:
  config: '{
      "cniVersion": "0.3.1",
      "name": "some-name",
      "type": "cni-name",
    }'
```

Kubevirt will eventually create a domain interface configuration that
points to the pod interface (e.g. `pod12345678`).
This is a snippet of the configuration:

```xml
<interface type='ethernet'>
   <alias name='ua-mynetwork'/>

>    <target dev='pod12345678' managed='no'/>

Reviewer 3: Please mention the target will be the tap name or the pod interface name if the tap doesn't exist.

   <model type='virtio-non-transitional'/>

>    <mac address='12:34:56:78:9a:bc'/>

Author: Done

   <mtu size='1500'/>
   <rom enabled='no'/>
</interface>
```

> A classic example of such a plugin is thehttps://github.com/kubevirt/kubevirt/blob/main/pkg/hooks/hooks.go#L31

Author: Done

[macvtap](https://kubevirt.io/user-guide/virtual_machines/net_binding_plugins/macvtap/)
plugin.

## The Sidecar Plugin

When a standard domain attachment requires customization,
or when additional services are needed (e.g. DHCP), a sidecar container
may be executed to integrate with Kubevirt.

The sidecar container runs in parallel to the virt-launcher container
which runs the hypervisor (libvirt and qemu).

> ### Sidecar Protocol

Reviewer: I think the content that related to Kubevirt sidecar hooks should be moved to dedicated file for that topic.
They we could reference it in this doc where needed.
I know there is no Kubevirt sidecar hook doc but what you wrote so far cover most of it.

Author: Perhaps, but it is not in my scope at the moment.
AFAIK the sidecar hook examples use a shim layer which in our context we do not use.

I have also not covered all APIs of the protocol, only the ones we found useful for our need.
By extracting this on its own, it will require a full coverage and explanations here that we do not need all.

Bottom line, for now I prefer to be focused only on the network binding plugin needs.

Kubevirt supports an integration hook between the virt-launcher
and sidecar containers that run in the same pod.

The integration is based on a client-server gRPC protocol.
The server side is setup at the sidecar container and the client
operates at the virt-launcher container.

From the sidecar perspective, the operation steps are as follows:

> - The sidecar container is defined as part of the virt-launcher pod,

Reviewer: nit:
In the following example there are bullets before each explanation, I think they look odd.

    vmi := &v1.VirtualMachineInstance{}
    if err := json.Unmarshal(params.GetVmi(), vmi); err != nil {
        return nil, fmt.Errorf("failed to unmarshal VMI: %v", err)
    }

    domainSpec := &domainschema.DomainSpec{
        // Unmarshalling domain spec makes the XML namespace attribute empty.
        // Some domain parameters requires namespace to be defined.
        // e.g: https://libvirt.org/drvqemu.html#pass-through-of-arbitrary-qemu-commands
        XmlNS: libvirtDomainQemuSchema,
    }
    if err := xml.Unmarshal(domainXML, domainSpec); err != nil {
        return nil, fmt.Errorf("failed to unmarshal given domain spec: %v", err)
    }

    updatedDomainSpec, err := changeDomain(domainSpec, vmi)
    if err != nil {
        return nil, err
    }

    newDomainXML, err := xml.Marshal(updatedDomainSpec)
    if err != nil {
        return nil, fmt.Errorf("failed to marshal updated domain spec: %v", err)
    }

    return &hooksv1alpha2.OnDefineDomainResult{
        DomainXML: newDomainXML,
    }, nil
}
```

It is recommended to use the latest available hook version.
The unused methods must be defined to satisfy the interface, preferably with
a default non-error response.

> #### Pod interface name

Reviewer: ```suggestion
#### Pod interface naming
```

Author: Done

The name of the network interface in the pod, to which the relevant network
is to bind to, is expected to be known.

- For the primary network, the default `eth0` naming is expected.

> - For the secondary networks, Kubevirt explicitly specifies the name through
>   Multus annotations of the pod. By that it assumes the CNI plugin used

Reviewer: ```suggestion
For the secondary networks, Kubevirt explicitly specifies the name through the pod's
  Multus network selection annotation. By that it assumes the CNI plugin used
```

Author: Done

>   confirms with the CNI specification and will name the interface accordingly.

Reviewer: > By that it assumes the CNI plugin used confirms with the CNI specification and will name the interface accordingly

I dont quite undrestand what does it means, could you please rephrase?

Author: Done

  The name used is generated by Kubevirt with the following format:
  `pod<network name hash ID>`.

>   - Network hash ID: The hash ID is generated based on the network name specified

Reviewer: nit: If you meant for the placeholder about, I think it should start with `Network name hash ID` (add 'name').

Author: Done

    in the VM/VMI interface & network spec.
    To programmatically generate it, one can use the following code snippet:

```go
import "kubevirt.io/kubevirt/pkg/network/namescheme"

ifaceName := namescheme.GenerateHashedInterfaceName(network_name)
```

#### Domain configuration

Changing the domain configuration usually involves the addition of a device
or a change in an existing device.

Either way, it is important to understand which are the networking
devices and what are the expectations from Kubevirt:

> - Domain API: "kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/api"

Reviewer: nit: I would add new line after each bullet in this section.

Author: Done

  - `DomainSpec`: The domain configuration spec structure, which is in sync
    with [libvirt domain configuration](https://libvirt.org/formatdomain.html).
    - `Devices`:
      - `Interface`: The network interface configuration spec.
      - `HostDevice`: The general host device configuration spec.
        The host-device is used to pass-through devices from the host to the domain.
        SR-IOV virtual functions are a classic example for networking.
- Device Alias:
  Kubevirt marks on the domain device spec an alias that references
  the logical network name as appearing in the VM/VMI spec.
  - For SR-IOV devices, the alias also includes a prefix: `sriov-`.

  ```go
  import domainschema "kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/api"
  deviceAlias := domainschema.NewUserDefinedAlias("sriov-" + "mynet1")
  ```
- Parameters: Several configuration parameters of an interface from the VMI spec
  are passed to the domain device configuration directly (e.g. the mac address).
  The domain interface and host-device expose a larger amount of parameters

>   which are not exposed through the VMI interface spec.

Reviewer: ```suggestion
  The domain interface and host-device may expose additional  parameters
  which are not exposed through the VMI interface spec.

Author: `may` is not a fit, but I did drop the "larger amount".

Done

  Plugin authors may populate other domain parameters if needed, taking
  the values as hard-coded or from the VMI object (including annotation).

### Sidecar Artifacts

The expected artifacts include:

- Documentation on how to register it to Kubevirt and what it does.
- A container image that includes the logic of a network binding plugin sidecar.

> - Optionally, a link to a registry that contains the container image.

Reviewer: I think these should go together, you provide a container image by persisting it in a registry and providing the  image tag that include the registry.

Author: I do not think a link to a registry is mandatory.

It is one thing to provide the sources, another to provide the raw image and yet another to provide it through a registry.

I will add a note about pushing to a registry in order to make it available for consumption in a cluster.
